### PR TITLE
feat: datasource check for WSL

### PR DESCRIPTION
**Pull request by CarlosNihelton**
## Proposed Commit Message

```
feat: datasource check for WSL

Adds `dscheck_WSL` function in ds-identify to allow detecting WSL.

Fixes: UDENG-1993
```

## Additional Context

This PR opens up the work for implementing a datasource for WSL.
Subsequent PR's will bring the datasource, its documentation and
the final step should be enabling it in the required datasources lists.

## Checklist

- [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/index.html)
- [x] I have updated or added any [unit tests](https://cloudinit.readthedocs.io/en/latest/development/testing.html) accordingly
- [ ] I have updated or added any [documentation](https://cloudinit.readthedocs.io/en/latest/development/contribute_docs.html) accordingly

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

File: tools/ds-identify
Comment (holmanb) [inline tools/ds-identify:1576]: Long options please (`grep --count`), for easier reading.

Also, matching on `'9p .*drvfs'` is still fairly broad - is there anything that we can do to tighten up this match?

File: tests/unittests/test_ds_identify.py
Comment (holmanb) [inline tests/unittests/test_ds_identify.py:2174]: Will the line in `/proc/mounts` always match `aname=drvfs`?

**Review by holmanb — COMMENTED** [review]
Thanks for this contribution @CarlosNihelton! See the changes requested below.

File: tools/ds-identify
Comment (dermotbradley) [reply tools/ds-identify:1576]: @holmanb Busybox's grep (as used by Alpine) only handles short options...

**Comment by dermotbradley** [comment]
Um, this PR changes ds-identify to recognise a WSL environment and to then use the WSL datasource...but there is *no* WSL DataSource in cloud-init and this PR does not introduce one...

File: tools/ds-identify
Comment (holmanb) [reply tools/ds-identify:1576]: @dermotbradley oh it does?

@CarlosNihelton I retract that part of the request, the other part still stands

File: tests/unittests/test_ds_identify.py
Comment (CarlosNihelton) [reply tests/unittests/test_ds_identify.py:2174]: Indeed it always match. The device name (the first field in `/proc/mounts`) changed in December (it used to be `drvfs` but it's now `LETTER:\\NUMBER`), thus we cannot rely on the device name. Yet across many versions of WSL I could check this part `aname=drvfs` is stable. `path=` is always there too, but I'd be skeptical about how stable that metadata position is to match it via regex in a single match.

I'd say `'^\S* \S* 9p \S*aname=drvfs;.*'` is a strong enough filter. That only matches entry where the third field is '9p' and the fourth contains 'aname=drvfs', where fields are separated by whitespaces.

File: tools/ds-identify
Comment (CarlosNihelton) [reply tools/ds-identify:1574]: The best I could find was a talk about the WSL2 architecture given by Microsoft's principal engineer and the product manager on YouTube:
https://youtu.be/lwhMThePdIo?t=2431&si=JKTHx39TyRgPbzkZ

File: tools/ds-identify
Comment (holmanb) [inline tools/ds-identify:1569]: WSL only supports Linux, right? Since ds-identify runs on other kernels besides just the linux kernel, lets add an early check to based on information we already know to avoid the grep on non-Linux.

```suggestion
    if [ ! "${DI_VIRT}" = "wsl" -o "$DI_UNAME_KERNEL_NAME" != "Linux" ]; then
```

cc @igalic

File: tools/ds-identify
Comment (holmanb) [inline tools/ds-identify:1579]: I don't think that `\S` is a [Posix regex](https://en.wikibooks.org/wiki/Regular_Expressions/POSIX_Basic_Regular_Expressions), which we prefer for portability.

From Gnu grep's `info` page I see:

```info
‘\S’
     Match non-whitespace, it is a synonym for ‘[^[:space:]]’.
```

So I think we want somthing more like:
```suggestion
    if [ "$(grep -c '^[^[:space:]]* [^[:space:]]* 9p [^[:space:]]*aname=drvfs;.*' "${PATH_ROOT}/proc/mo  unts")" = 0 ]; then
```

Which also works in Gnu and busybox (cc: @dermotbradley) grep implementations:
```bash
$ echo "C:\\134 /mnt/c 9p rw,dirsync,aname=drvfs;path=C:\\;uid=0;" | grep -c '^[^[:space:]]* [^[:space:]]* 9p [^[:space:]]*aname=drvfs;.*'
1
$ echo "C:\\134 /mnt/c 9p rw,dirsync,aname=drvfs;path=C:\\;uid=0;" | busybox grep -c '^[^[:space:]]* [^[:space:]]* 9p [^[:space:]]*aname=drvfs;.*'
1
```

**Review by holmanb — CHANGES_REQUESTED** [review]
@CarlosNihelton Looking better, thanks for the updates. I have a couple more requested changes.

File: tools/ds-identify
Comment (igalic) [reply tools/ds-identify:1569]: yeah, if we don't, everything else will also needlessly fail

Comment (igalic) [reply tools/ds-identify:1569]: the l in wsl stands for Linux, so no other platform should match that, to begin with

**Review by blackboxsw — CHANGES_REQUESTED** [review]
Thanks for this branch @CarlosNihelton and @holmanb for the review here. A note that we should not merge this branch until we also have DataSourceWSL.py defined and merged. Normally this is all done in one giant PR, python datasouce + ds-identify implementation, but I get that separating the PR into two unique PRs may make this review easier to review/land. No need to include DataSourceWSL.py implementation if not desired, we just need to wait to merge this PR until we are certain DataSourceWSL exists.

The reason we can't land a ds-identify only branch before the matching DataSourceWSL.py implementation because if ds-identity does actually detect WSL, it can't actually choose that datasource because it doesn't yet exist in the python module `cloudinit.sources.DatasourceWSL`

If ds-identify does detect the possibility of WSL during systemd generator timeframe, the python code of cloud-init will emit and error like the following:
```
root@nnn:~# cloud-init status --long
status: done
extended_status: degraded done
boot_status_code: enabled-by-generator
last_update: Mon, 08 Jan 2024 20:37:18 +0000
detail:
DataSourceLXD
errors: []
recoverable_errors:
ERROR:
	- Could not import DataSourceWSL. Does the DataSource exist and is it importable?
```


Marking 'request changes' just to track on objection to merging until we land the prerequisite PR delivering cloudinit/sources/DataSourceWSL.py

**Comment by holmanb** [comment]
@blackboxsw I agree that typically we wouldn't want to see a new datasource land until it is in a working state. However, this is an exception from the norm for a number of reasons: 1) the upstream devs had an in-person meeting with those implementing the datasource 2) they have a shared their implementation plan and gotten agreement from us 3) we know that engineering time is allocated to this effort. Compare this to PRs where an unknown dev from a cloud provider reaches out for code inclusion without any prior contact or expectations set.

> If ds-identify does detect the possibility of WSL during systemd generator timeframe, the python code of cloud-init will emit and error like the following:

This is only a hypothetical scenario, not a failure path that should be seen on any supported version of cloud-init. A positive identification of WSL must only happen on WSL. If we merge this and someone tries running cloud-init on WSL - it isn't a supported platform of cloud-init (yet) so what is the harm? Additionally, to get to the failure path that you are describing, one would have to either 1) include `WSL` in the `datasource_list` of their image or 2) use the built-in list and detect no other datasource (WSL is last on the built-in list). Both of these are unsupported scenarios - both before and after we merge this PR.

If we merge this and this gets detected on a not-WSL platform, then that is a bug in this implementation of `dscheck_WSL` which would exist whether or not there is a Python datasource implemented. If we merge this before the Python datasource implementation, we could get daily tests running sooner and get an earlier detection of bugs like that.

> if ds-identity does actually detect WSL, it can't actually choose that datasource because it doesn't yet exist in the python module cloudinit.sources.DatasourceWSL

Not wrong, but what is to stop us from implementing this in parts? WSL support would not be "complete", but this wouldn't break any existing datasources.

> And I think we want to wait to land this particular PR until we have an implementation of DataSourceWSL.py merged.

@blackboxsw I'm fine with waiting on this PR if you really want to. In my opinion, however, the primary reason to do this is to keep a standard upstream process for all datasource proposals. The opposing argument is that if we go ahead with a merge we can avoid this PR bitrotting and also get sooner detection of bugs from daily testing. Since the effort to get a `WSL` datasource merged is targeting the 24.04 release of Ubuntu, I think that this is a pretty low risk PR to merge.

**Review by holmanb — APPROVED** [review]
Great work on this PR @CarlosNihelton, thanks! This looks ready to me, but you'll need to wait on the concerns of @blackboxsw before this can merge.

**Comment by blackboxsw** [comment]
@holmanb your points are well made. There really isn't a significant concern about a "recoverable error" message in cloud-init on WSL images booting with tip of cloud-init because we, today, have no support whatsoever for WSL images, so it's a minor step toward the end goal and allows the WSL devs to start their local testing during initial datasource development.

And since we have a priority commitment from the devs for WSL datasource support. There is little harm in cloud-init trying to detect WSL environments and emitting a warning/error that WSL support is not yet present.

Let's not let me block this since the work is active and making progress. Thanks for defending good @CarlosNihelton's work :)

File: tools/ds-identify
Comment (blackboxsw) [inline tools/ds-identify:1569]: I'm pretty sure all distros we support run ds-identify from POSIX-compliant implementations. But, there is an off chance that a minimal image may have a busybox implementation somewhere that doesn't adhere strictly to POSIX order of operators here.
In this case and below, to be clear in all cases, let's just use the `!=` operator to be safe.
```suggestion
    if [ "${DI_UNAME_KERNEL_NAME}" != "Linux" ]; then
```

File: tools/ds-identify
Comment (blackboxsw) [inline tools/ds-identify:1573]: Safety for potential non-POSIX compliant images
```suggestion
    if [ "${DI_VIRT}" != "wsl" ]; then
```

**Review by blackboxsw — CHANGES_REQUESTED** [review]
Minor suggestion `just in case`™

**Review by blackboxsw — APPROVED** [review]
+! LGTM!

**Pull request merged**
